File: src/test/scala/com/exasol/cloudetl/orc/converter/OrcConverterPrimitiveTypesTest.scala
package com.exasol.cloudetl.orc.converter

import org.apache.hadoop.hive.common.`type`.HiveDecimal
import org.apache.hadoop.hive.ql.exec.vector._
import org.apache.orc.TypeDescription._

@SuppressWarnings(Array("org.wartremover.warts.AsInstanceOf"))
class OrcConverterPrimitiveTypesTest extends BaseOrcConverterTest {

  test("reads BOOLEAN as long value") {
    val schema = createStruct().addField("boolean", createBoolean())
    orcWriter.write[Any](schema, List(true, false, null))
    val records = getRecords()
    assert(records(0).get(0) === true)
    assert(records(1).get(0) === false)
    assert(records(2).isNullAt(0) === true)
  }

  test("reads BYTE as byte value") {
    val schema = createStruct().addField("byte", createByte())
    orcWriter.write[Any](schema, List(13, null))
    val records = getRecords()
    assert(records(0).get(0) === 13)
    assert(records(1).isNullAt(0) === true)
  }

  test("reads SHORT as short value") {
    val schema = createStruct().addField("short", createShort())
    orcWriter.write[Any](schema, List(314, null))
    val records = getRecords()
    assert(records(0).get(0) === 314)
    assert(records(1).isNullAt(0) === true)
  }

  test("reads INT as integer value") {
    val schema = createStruct().addField("int", createInt())
    orcWriter.write[Any](schema, List(314, null))
    val records = getRecords()
    assert(records(0).get(0) === 314)
    assert(records(1).isNullAt(0) === true)
  }

  test("reads LONG as long value") {
    val schema = createStruct().addField("long", createLong())
    orcWriter.write[Any](schema, List(3L, null))
    val records = getRecords()
    assert(records(0).get(0) === 3L)
    assert(records(1).isNullAt(0) === true)
  }

  test("reads FLOAT as float value") {
    val schema = createStruct().addField("float", createFloat())
    orcWriter.write[Any](schema, List(3.14F, null))
    val records = getRecords()
    assert(records(0).get(0) === 3.14f)
    assert(records(1).isNullAt(0) === true)
  }

  test("reads DOUBLE as double value") {
    val schema = createStruct().addField("double", createDouble())
    orcWriter.write[Any](schema, List(2.71, null))
    val records = getRecords()
    assert(records(0).get(0) === 2.71)
    assert(records(1).isNullAt(0) === true)
  }

  test("reads Decimal value as java.math.decimal") {
    val schema = createStruct().addField("decimal", createDecimal())
    val batch = schema.createRowBatch()
    batch.size = 2
    withWriter(schema) { writer =>
      val decimalVector = batch.cols(0).asInstanceOf[DecimalColumnVector]
      decimalVector.noNulls = false
      decimalVector.vector(0).set(HiveDecimal.create("173.433"))
      decimalVector.isNull(1) = true
      writer.addRowBatch(batch)
    }
    val records = getRecords()
    assert(records(0).get(0).isInstanceOf[java.math.BigDecimal])
    assert(records(0).getAs[java.math.BigDecimal](0).doubleValue() === 173.433)
    assert(records(1).isNullAt(0) === true)
  }

  test("reads DATE as java.sql.date value") {
    val schema = createStruct().addField("date", createDate())
    orcWriter.write[Any](schema, List(0, 1, null))
    val records = getRecords()
    assert(records(0).get(0) === java.sql.Date.valueOf("1970-01-01"))
    assert(records(1).get(0) === java.sql.Date.valueOf("1970-01-02"))
    assert(records(2).isNullAt(0) === true)
  }

  test("reads TIMESTAMP as java.sql.timestamp value") {
    val timestamp1 = java.sql.Timestamp.from(java.time.Instant.EPOCH)
    val timestamp2 = new java.sql.Timestamp(System.currentTimeMillis())
    val schema = createStruct().addField("timestamp", createTimestamp())
    orcWriter.write[Any](schema, List(timestamp1, timestamp2, null))
    val records = getRecords()
    assert(records(0).get(0) === timestamp1)
    assert(records(1).get(0) === timestamp2)
    assert(records(2).isNullAt(0) === true)
  }

  test("reads CHAR as string value") {
    val schema = createStruct().addField("string", createChar())
    orcWriter.write[Any](schema, List("value", null))
    val records = getRecords()
    assert(records(0).get(0) === "value")
    assert(records(1).isNullAt(0) === true)
  }

  // scalastyle:off nonascii
  test("reads STRING as string value") {
    val schema = createStruct().addField("string", createString())
    orcWriter.write[Any](schema, List("välue", null))
    val records = getRecords()
    assert(records(0).get(0) === "välue")
    assert(records(1).isNullAt(0) === true)
  }

  test("reads VARCHAR as string value") {
    val schema = createStruct().addField("string", createVarchar())
    orcWriter.write[Any](schema, List("smiley ☺", null))
    val records = getRecords()
    assert(records(0).get(0) === "smiley ☺")
    assert(records(1).isNullAt(0) === true)
  }
  // scalastyle:on nonascii

  test("reads BINARY as string value") {
    val schema = createStruct().addField("string", createBinary())
    orcWriter.write[Any](schema, List("str", null))
    val records = getRecords()
    assert(records(0).get(0) === "str")
    assert(records(1).isNullAt(0) === true)
  }

}

File: src/test/java/com/exasol/cloudetl/orc/converter/OrcConverterPrimitiveTypesTest.java
package com.exasol.cloudetl.orc.converter;

import static org.apache.orc.TypeDescription.*;
import static org.junit.jupiter.api.Assertions.*;

import java.util.Arrays;

import org.junit.jupiter.api.Test;

class OrcConverterPrimitiveTypesTest extends BaseOrcConverterTest {
    @Test
    void readsBooleanAsLongValue() throws Exception {
        write(createStruct().addField("boolean", createBoolean()), Arrays.asList(true, false, null));
        final var records = getRecords();
        assertEquals(true, records.get(0).get(0));
        assertEquals(false, records.get(1).get(0));
        assertTrue(records.get(2).isNullAt(0));
    }

    @Test
    void readsByteAsByteValue() throws Exception {
        write(createStruct().addField("byte", createByte()), Arrays.asList(13, null));
        assertEquals((byte) 13, getRecords().get(0).get(0));
        assertTrue(getRecords().get(1).isNullAt(0));
    }

    @Test
    void readsShortAsShortValue() throws Exception {
        write(createStruct().addField("short", createShort()), Arrays.asList(314, null));
        assertEquals((short) 314, getRecords().get(0).get(0));
        assertTrue(getRecords().get(1).isNullAt(0));
    }

    @Test
    void readsIntAsIntegerValue() throws Exception {
        write(createStruct().addField("int", createInt()), Arrays.asList(314, null));
        assertEquals(314, getRecords().get(0).get(0));
        assertTrue(getRecords().get(1).isNullAt(0));
    }

    @Test
    void readsLongAsLongValue() throws Exception {
        write(createStruct().addField("long", createLong()), Arrays.asList(3L, null));
        assertEquals(3L, getRecords().get(0).get(0));
        assertTrue(getRecords().get(1).isNullAt(0));
    }

    @Test
    void readsFloatAsFloatValue() throws Exception {
        write(createStruct().addField("float", createFloat()), Arrays.asList(3.14f, null));
        assertEquals(3.14f, getRecords().get(0).get(0));
        assertTrue(getRecords().get(1).isNullAt(0));
    }

    @Test
    void readsDoubleAsDoubleValue() throws Exception {
        write(createStruct().addField("double", createDouble()), Arrays.asList(2.71, null));
        assertEquals(2.71, getRecords().get(0).get(0));
        assertTrue(getRecords().get(1).isNullAt(0));
    }

    @Test
    void readsDecimalValueAsJavaMathDecimal() throws Exception {
        write(createStruct().addField("decimal", createDecimal()), Arrays.asList("173.433", null));
        final var records = getRecords();
        assertInstanceOf(java.math.BigDecimal.class, records.get(0).get(0));
        assertEquals(173.433, ((java.math.BigDecimal) records.get(0).get(0)).doubleValue());
        assertTrue(records.get(1).isNullAt(0));
    }

    @Test
    void readsDateAsJavaSqlDateValue() throws Exception {
        write(createStruct().addField("date", createDate()), Arrays.asList(0, 1, null));
        final var records = getRecords();
        assertEquals(java.sql.Date.valueOf("1970-01-01"), records.get(0).get(0));
        assertEquals(java.sql.Date.valueOf("1970-01-02"), records.get(1).get(0));
        assertTrue(records.get(2).isNullAt(0));
    }

    @Test
    void readsTimestampAsJavaSqlTimestampValue() throws Exception {
        final java.sql.Timestamp timestamp1 = java.sql.Timestamp.from(java.time.Instant.EPOCH);
        final java.sql.Timestamp timestamp2 = new java.sql.Timestamp(System.currentTimeMillis());
        write(createStruct().addField("timestamp", createTimestamp()), Arrays.asList(timestamp1, timestamp2, null));
        final var records = getRecords();
        assertEquals(timestamp1, records.get(0).get(0));
        assertEquals(timestamp2, records.get(1).get(0));
        assertTrue(records.get(2).isNullAt(0));
    }

    @Test
    void readsCharAsStringValue() throws Exception {
        write(createStruct().addField("string", createChar()), Arrays.asList("value", null));
        assertEquals("value", getRecords().get(0).get(0));
        assertTrue(getRecords().get(1).isNullAt(0));
    }

    @Test
    void readsStringAsStringValue() throws Exception {
        write(createStruct().addField("string", createString()), Arrays.asList("välue", null));
        assertEquals("välue", getRecords().get(0).get(0));
        assertTrue(getRecords().get(1).isNullAt(0));
    }

    @Test
    void readsVarcharAsStringValue() throws Exception {
        write(createStruct().addField("string", createVarchar()), Arrays.asList("smiley ☺", null));
        assertEquals("smiley ☺", getRecords().get(0).get(0));
        assertTrue(getRecords().get(1).isNullAt(0));
    }

    @Test
    void readsBinaryAsStringValue() throws Exception {
        write(createStruct().addField("string", createBinary()), Arrays.asList("str", null));
        assertEquals("str", getRecords().get(0).get(0));
        assertTrue(getRecords().get(1).isNullAt(0));
    }
}
